File: src/lib/upHeroNarrative.ts
/**
 * Up Hero — Phase 3: 전투 narrative 생성기.
 *
 * 템플릿:
 *  - 영웅 공격: `{critPrefix}영웅이 {monster.name}의 {bodyPart}를 {verb}. −{dmg}`
 *  - 몬스터 공격: `{critPrefix}{monster.name}이(가) {instrument} 영웅을 {verb}. −{dmg}`
 *  - 회피/미스: 단일 문장 풀에서 pick
 *
 * `critPrefix` = `"치명타! "` 또는 빈 문자열.
 *
 * 순수 함수 — same outcome + damage 조합이어도 pool 에서 random pick 하므로 매번 다른 문장 생성.
 * CombatSession.log 에 저장할 때는 1회 호출 결과를 persist 해서 재렌더 시 일관성 유지.
 *
 * Phase 13c — i18n. 기존 단일 string 반환 외에 `heroAttackNarrativeI18n` /
 *   `monsterAttackNarrativeI18n` 이 { text, key, params } 를 반환해 LogEntry
 *   에 저장. CombatLog 는 key + params 를 현재 언어로 풀어 렌더. 한국어 fallback
 *   은 legacy save 호환 + i18n 키 누락 시 안전망.
 */

import type { CombatOutcome, Monster, NarrativeParams } from "@/types/uphero";
import {
  HERO_HIT_VERBS,
  HERO_CRIT_VERBS,
  HERO_MISS_LINES,
  HERO_DODGE_LINES,
  MONSTER_ATTACK_FLAVOR,
  MONSTER_BODY_PARTS,
  MONSTER_DODGE_LINES,
  MONSTER_MISS_LINES,
} from "@/data/upHeroCombatFlavor";

/** 배열에서 random 요소 pick */
function pick<T>(pool: readonly T[]): T {
  return pool[Math.floor(Math.random() * pool.length)];
}

/** Phase 13c — i18n metadata 와 함께 묶인 narrative. */
export interface I18nNarrative {
  /** 한국어 legacy fallback */
  text: string;
  /** i18n key (없으면 text 그대로 렌더) */
  key?: string;
  /** key 를 t() 에 전달할 때 쓸 params. monsterTemplateId 포함 가능. */
  params?: NarrativeParams;
}

/**
 * 영웅이 몬스터를 공격 — outcome + damage → narrative 생성.
 *
 * @param monster 대상 몬스터 (이름, kind 사용)
 * @param outcome "hit" | "crit" | "dodge" | "miss"
 *                "dodge" 면 방어자(몬스터) 가 피함
 *                "miss" 면 공격자(영웅) 가 허탕침
 * @param damage 가한 피해 (dodge/miss 면 0)
 */
export function heroAttackNarrative(
  monster: Monster,
  outcome: CombatOutcome,
  damage: number,
): string {
  return heroAttackNarrativeI18n(monster, outcome, damage).text;
}

/**
 * Phase 13c — i18n-ready variant. 한국어 text + i18n key + params 반환.
 */
export function heroAttackNarrativeI18n(
  monster: Monster,
  outcome: CombatOutcome,
  damage: number,
): I18nNarrative {
  switch (outcome) {
    case "miss":
      return { text: pick(HERO_MISS_LINES), key: "uphero.combat.narrative.heroMiss" };
    case "dodge":
      return {
        text: pick(MONSTER_DODGE_LINES[monster.kind]),
        key: "uphero.combat.narrative.monsterDodge",
        params: {
          monster: monster.name,
          monsterTemplateId: monster.templateId ?? "",
        },
      };
    case "crit": {
      const part = pick(MONSTER_BODY_PARTS[monster.kind].weak);
      const verb = pick(HERO_CRIT_VERBS);
      return {
        text: `치명타! 영웅이 ${monster.name}의 ${part}를 ${verb}. −${damage}`,
        key: "uphero.combat.narrative.heroCrit",
        params: {
          monster: monster.name,
          monsterTemplateId: monster.templateId ?? "",
          damage,
        },
      };
    }
    case "hit": {
      const part = pick(MONSTER_BODY_PARTS[monster.kind].normal);
      const verb = pick(HERO_HIT_VERBS);
      return {
        text: `영웅이 ${monster.name}의 ${part}를 ${verb}. −${damage}`,
        key: "uphero.combat.narrative.heroHit",
        params: {
          monster: monster.name,
          monsterTemplateId: monster.templateId ?? "",
          damage,
        },
      };
    }
  }
}

/**
 * 몬스터가 영웅을 공격 — outcome + damage → narrative 생성.
 *
 * @param monster 공격 중인 몬스터 (이름, kind 사용)
 * @param outcome "hit" | "crit" | "dodge" | "miss"
 *                "dodge" 면 방어자(영웅) 가 피함
 *                "miss" 면 공격자(몬스터) 가 허탕침
 * @param damage 가한 피해 (dodge/miss 면 0)
 */
export function monsterAttackNarrative(
  monster: Monster,
  outcome: CombatOutcome,
  damage: number,
): string {
  return monsterAttackNarrativeI18n(monster, outcome, damage).text;
}

/**
 * Phase 13c — i18n-ready variant. 한국어 text + i18n key + params 반환.
 */
export function monsterAttackNarrativeI18n(
  monster: Monster,
  outcome: CombatOutcome,
  damage: number,
): I18nNarrative {
  const flavor = MONSTER_ATTACK_FLAVOR[monster.kind];
  switch (outcome) {
    case "miss":
      return {
        text: pick(MONSTER_MISS_LINES[monster.kind]),
        key: "uphero.combat.narrative.monsterMiss",
        params: {
          monster: monster.name,
          monsterTemplateId: monster.templateId ?? "",
        },
      };
    case "dodge":
      return { text: pick(HERO_DODGE_LINES), key: "uphero.combat.narrative.heroDodge" };
    case "crit": {
      const verb = pick(flavor.critVerbs);
      return {
        text: `치명타! ${monster.name}이(가) ${verb}. −${damage}`,
        key: "uphero.combat.narrative.enemyCrit",
        params: {
          monster: monster.name,
          monsterTemplateId: monster.templateId ?? "",
          damage,
        },
      };
    }
    case "hit": {
      const instrument = pick(flavor.instruments);
      const verb = pick(flavor.hitVerbs);
      return {
        text: `${monster.name}이(가) ${instrument} 영웅을 ${verb}. −${damage}`,
        key: "uphero.combat.narrative.enemyHit",
        params: {
          monster: monster.name,
          monsterTemplateId: monster.templateId ?? "",
          damage,
        },
      };
    }
  }
}

File: upnext-ios/UpNext/UpNext/Models/UpHeroNarrative.swift
//
//  UpHeroNarrative.swift
//  UpNext 로직 — Up Hero 전투 narrative 생성기.
//
//  웹 src/lib/upHeroNarrative.ts (178줄) 1:1 포팅.
//  Phase 2.4 (RPG 엔진) 오케스트레이션 데이터 레이어 산출물.
//
//  outcome + damage → narrative 문장. pick 은 시드 가능한 RandomSource 사용.
//

import Foundation

/// i18n metadata 와 함께 묶인 narrative. 웹 `I18nNarrative`.
struct I18nNarrative: Equatable {
    let text: String                   // 한국어 legacy fallback
    var key: String? = nil             // i18n key
    var params: NarrativeParams? = nil  // t() 전달용 params
}

enum UpHeroNarrative {

    /// 배열에서 random 요소 pick. 웹 `pick` (seedable rng).
    static func pick<T, R: RandomSource>(_ pool: [T], rng: inout R) -> T {
        pool[rng.int(below: pool.count)]
    }

    /// narrativeKey + params 를 인앱 언어로 해석 — 전투 로그 현지화 chokepoint.
    /// 카탈로그의 `{param}` 토큰을 값으로 치환(언어별 어순 무관). monster·dungeon·
    /// description·equipment·who 등 콘텐츠 파라미터는 카탈로그 경유 재현지화(locRuntime).
    /// 키가 카탈로그에 없으면 한국어 fallback 반환(안전).
    static func resolveLog(_ key: String, _ params: NarrativeParams?, fallback: String) -> String {
        let template = AppConfig.loc(String.LocalizationValue(key))
        if template == key { return fallback }   // 미등록 키 → 한국어 fallback
        guard let params, !params.isEmpty else { return template }
        var s = template
        let contentParams: Set<String> = ["monster", "dungeon", "description", "equipment", "who"]
        for (name, val) in params {
            let rep: String
            switch val {
            case .text(let t):
                rep = contentParams.contains(name) ? AppConfig.locRuntime(t) : t
            case .number(let n):
                rep = String(Int(n.rounded()))
            }
            s = s.replacingOccurrences(of: "{\(name)}", with: rep)
        }
        return s
    }

    /// 몬스터 이름/templateId params.
    private static func monsterParams(_ m: Monster) -> NarrativeParams {
        ["monster": .text(m.name), "monsterTemplateId": .text(m.templateId ?? "")]
    }

    /// 몬스터 params + damage.
    private static func monsterParamsD(_ m: Monster, _ damage: Int) -> NarrativeParams {
        var p = monsterParams(m)
        p["damage"] = .number(Double(damage))
        return p
    }

    /// 영웅 공격 narrative (i18n). 웹 `heroAttackNarrativeI18n`.
    static func heroAttackNarrativeI18n<R: RandomSource>(
        monster: Monster, outcome: CombatOutcome, damage: Int, rng: inout R
    ) -> I18nNarrative {
        switch outcome {
        case .miss:
            return I18nNarrative(text: pick(CombatFlavor.heroMissLines, rng: &rng),
                                 key: "uphero.combat.narrative.heroMiss")
        case .dodge:
            return I18nNarrative(
                text: pick(CombatFlavor.monsterDodgeLines[monster.kind]!, rng: &rng),
                key: "uphero.combat.narrative.monsterDodge",
                params: monsterParams(monster))
        case .crit:
            let part = pick(CombatFlavor.monsterBodyParts[monster.kind]!.weak, rng: &rng)
            let verb = pick(CombatFlavor.heroCritVerbs, rng: &rng)
            return I18nNarrative(
                text: "치명타! 영웅이 \(monster.name)의 \(part)를 \(verb). −\(damage)",
                key: "uphero.combat.narrative.heroCrit",
                params: monsterParamsD(monster, damage))
        case .hit:
            let part = pick(CombatFlavor.monsterBodyParts[monster.kind]!.normal, rng: &rng)
            let verb = pick(CombatFlavor.heroHitVerbs, rng: &rng)
            return I18nNarrative(
                text: "영웅이 \(monster.name)의 \(part)를 \(verb). −\(damage)",
                key: "uphero.combat.narrative.heroHit",
                params: monsterParamsD(monster, damage))
        }
    }

    /// 몬스터 공격 narrative (i18n). 웹 `monsterAttackNarrativeI18n`.
    static func monsterAttackNarrativeI18n<R: RandomSource>(
        monster: Monster, outcome: CombatOutcome, damage: Int, rng: inout R
    ) -> I18nNarrative {
        let flavor = CombatFlavor.monsterAttackFlavor[monster.kind]!
        switch outcome {
        case .miss:
            return I18nNarrative(
                text: pick(CombatFlavor.monsterMissLines[monster.kind]!, rng: &rng),
                key: "uphero.combat.narrative.monsterMiss",
                params: monsterParams(monster))
        case .dodge:
            return I18nNarrative(text: pick(CombatFlavor.heroDodgeLines, rng: &rng),
                                 key: "uphero.combat.narrative.heroDodge")
        case .crit:
            let verb = pick(flavor.critVerbs, rng: &rng)
            return I18nNarrative(
                text: "치명타! \(monster.name)이(가) \(verb). −\(damage)",
                key: "uphero.combat.narrative.enemyCrit",
                params: monsterParamsD(monster, damage))
        case .hit:
            let instrument = pick(flavor.instruments, rng: &rng)
            let verb = pick(flavor.hitVerbs, rng: &rng)
            return I18nNarrative(
                text: "\(monster.name)이(가) \(instrument) 영웅을 \(verb). −\(damage)",
                key: "uphero.combat.narrative.enemyHit",
                params: monsterParamsD(monster, damage))
        }
    }

    /// 영웅 공격 narrative (string). 웹 `heroAttackNarrative`.
    static func heroAttackNarrative<R: RandomSource>(
        monster: Monster, outcome: CombatOutcome, damage: Int, rng: inout R
    ) -> String {
        heroAttackNarrativeI18n(monster: monster, outcome: outcome, damage: damage, rng: &rng).text
    }

    /// 몬스터 공격 narrative (string). 웹 `monsterAttackNarrative`.
    static func monsterAttackNarrative<R: RandomSource>(
        monster: Monster, outcome: CombatOutcome, damage: Int, rng: inout R
    ) -> String {
        monsterAttackNarrativeI18n(monster: monster, outcome: outcome, damage: damage, rng: &rng).text
    }
}
